src/librustc/traits/error_reporting.rs

```diff
@@ -224,13 +222,30 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
                 }
             }
 
+
+            let span = obligation.cause.span;
+            let lint = self.get_lint_from_cause_code(&obligation.cause.code);
+            macro_rules! struct_span_err_or_lint {
```

Comment:
Nit: I feel like I've seen this macro at least twice. Can't we abstract this somehow? Or move it?

Reply:
It specifically takes advantage of variables in scope, so sadly it wouldn't be that easy.
We'd have to take both `infcx` and `obligation.cause` (instead of `tcx.sess` and `span`), I think? Maybe it *is* easy.

Reply:
What do you think `struct_obligation_err_or_lint`, as a name for the macro?
Also, is it just me or is `struct` just not the best word to use?
I almost want to move everything to diagnostic builders so we can get rid of the non-`struct_` versions and remove the `struct_` prefix.